include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -0,0 +1,301 @@
+#ifndef _NBL_BUILTIN_HLSL_CPP_COMPAT_IMPL_INTRINSICS_IMPL_INCLUDED_
+#define _NBL_BUILTIN_HLSL_CPP_COMPAT_IMPL_INTRINSICS_IMPL_INCLUDED_
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace cpp_compat_intrinsics_impl
+{
+template<typename T>
+struct dot_helper
+{
+	using scalar_type = typename vector_traits<T>::scalar_type;
+
+	static inline scalar_type dot(NBL_CONST_REF_ARG(T) lhs, NBL_CONST_REF_ARG(T) rhs)
+	{
+		static array_get<T, scalar_type> getter;
+		scalar_type retval = getter(lhs, 0) * getter(rhs, 0);
+
+		static const uint32_t ArrayDim = sizeof(T) / sizeof(scalar_type);
+		for (uint32_t i = 1; i < ArrayDim; ++i)
+			retval = retval + getter(lhs, i) * getter(rhs, i);
+
+		return retval;
+	}
+};
+
+#define DEFINE_BUILTIN_VECTOR_SPECIALIZATION(FLOAT_TYPE, RETURN_VALUE)\
+template<uint32_t N>\
+struct dot_helper<vector<FLOAT_TYPE, N> >\
+{\
+	using VectorType = vector<FLOAT_TYPE, N>;\
+	using ScalarType = typename vector_traits<VectorType>::scalar_type;\
+\
+	static inline ScalarType dot(NBL_CONST_REF_ARG(VectorType) lhs, NBL_CONST_REF_ARG(VectorType) rhs)\
+	{\
+		return RETURN_VALUE;\
+	}\
+};\
+
+#ifdef __HLSL_VERSION
+#define BUILTIN_VECTOR_SPECIALIZATION_RET_VAL dot(lhs, rhs)
```

Comment:
https://godbolt.org/z/7EPsWvPe7

```
DXC (trunk)
1
2
3
DXC (trunk) - 325ms

<source>:7:5: error: recursive functions are not allowed: function 'main' calls recursive function 'Functor<vector<float, 3> >::dot'
    static inline float dot(const in T lhs, const in T rhs)
    ^
```

only happens upon instantiation, but `::` before the HLSL intrinsics doesn't work, so you need to call your static function `__call` not `dot`

Reply:
I;ve filed a bug btw, https://github.com/microsoft/DirectXShaderCompiler/issues/7035